### DISCUSSION TITLE
How to handle the renaming of a project specific module?

### DISCUSSION THREAD

**Discussion by jcdrubay** [OCA/OpenUpgrade#3245] [General]
Today, it is needed to change the code of the apriori.py in the OpenUpgrade repo which encourages to do more project specific work in OpenUpgrade repo which will also discourage contributions back to OCA.

A nice way to handle this case should probably be documented in
https://oca.github.io/OpenUpgrade/devfaq.html

## Below is an answer from @legalsylvain on this topic initially reported here: https://github.com/jcdrubay/OpenUpgrade/issues/3

I faced the same question several times in the past. Each time I solved with a custom branch of openupgrade with a commit adding custom GRAP changes. Not the better, but it works.

If we want to allow the possibility to add custom rename / merge it should not be done in a pre-migration script of a custom module in my opinion. Two reasons : 
- the apriori.py is used both time. Once in ``openupgrade_framework``, once in ``upgrade_analysis``. If custom values are set in a custom modules, analysis will be wrong.
- in the openupgrade process the rename_modules / merge_modules is used at the very beginning of the upgrade (in the pre-migration.py script of the base module). AFAIK, at the end of the upgrade of base, the modules tree is recompputed (=order of the modules to update). If your custom modules is not renamed at this step, openupgrade will ignore it because it will not find the ``komit_base`` module in your addons path.

So for me, the single solution I see is to add in the ``apriori.py`` file, something like : 

```
_rename_modules = {
    # Odoo
    "crm_iap_lead": "crm_iap_mine",
    # ....
}

def _get_custom_rename_modules():
    # read the odoo.cfg file and search for a openupgrade_rename_modules section
    custom_conf = odoo.tools.config.misc.get('openupgrade_rename_modules', {})    
    return {x:y for x,y in custom_conf.items()}

rename_modules = _rename_modules + _get_custom_rename_modules()
```

And in the ``odoo.cfg`` modules 

```
[openupgrade_rename_modules]
audit_base=k3_audit_base
...
[openupgrade_merged_modules]
```

**Comment by jcdrubay** [comment]
I have started to work on this and it is starting to work.

Usually the file odoo.cfg is not committed (at least for us). But it would be good to commit the information related to custom openupgrade_renamed_modules and openupgrade_merged_modules.

I am planning to read a new configuration `openupgrade-config-path` from the odoo.cfg configuration file.

Then, a file stored at openupgrade-config-path will be added to the repository of the project and could contain all the specific custom configurations for `renamed_modules` and `merged_modules`.

Please, let me know you thoughts.

**Comment by legalsylvain** [comment]
Could you show the code ? Or better to make a PR ?

**Comment by jcdrubay** [comment]
I did exactly as you suggested @legalsylvain .
I will create a PR soon, I am still testing.

**Comment by jcdrubay** [comment]
Here is the code I am adding to load the custom config

```diff
+++ b/openupgrade_scripts/apriori.py
@@ -1,9 +1,42 @@
+import logging
+
+import configparser as ConfigParser
+
+import odoo
+
+def load_custom_config():
+    openupgrade_config_path = odoo.tools.config.misc.get("openupgrade").get("config_path")
+    res = {}
+    if openupgrade_config_path:
+        logging.info("load_custom_config from %s: ", openupgrade_config_path)
+        p = ConfigParser.RawConfigParser()
+        try:
+            p.read([openupgrade_config_path])
+            for sec in p.sections():
+                res.setdefault(sec, {})
+                for (name, value) in p.items(sec):
+                    if value=='True' or value=='true':
+                        value = True
+                    if value=='False' or value=='false':
+                        value = False
+                    res[sec][name] = value
+        except IOError:
+            pass
+        except ConfigParser.NoSectionError:
+            pass
+    else:
+        logging.info("load_custom_config, openupgrade_config_path not set")
+
+    return res
+
+custom_config = load_custom_config()
+
```

and then something like this:

```diff
+def _get_custom_rename_modules():
+    renamed_modules = custom_config.get("renamed_modules", {})
+    return renamed_modules

-}
+renamed_modules = _renamed_modules | _get_custom_rename_modules()
```

I will create a clean PR later today.

**Comment by jcdrubay** [comment]
Here is the PR:
https://github.com/OCA/OpenUpgrade/pull/3264/files

**Comment by jcdrubay** [comment]
Related discussion https://github.com/OCA/OpenUpgrade/discussions/3231
